README.md
gooser
======
gooser is a small GRPC user api written in golang. It provides:
* functions for creating, updating, deleting users & groups
* functions for resetting the password
* groups can have roles assigned

# settings
All settings have to be provided by environment variables:

| environment   variable         | description                                                                                                                                        | default                                |
|--------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------|
| GOOSER_ADMIN_USER              | A user with the given username will be created if it does not exist. The   user will be put in a group called "admins", having the "admin" role.   | admin                                  |
| GOOSER_CONFIRM_URL             | Base url which will be sent for confirming the user's mail address                                                                                 | http://localhost:1234/#/confirm-mail   |
| GOOSER_DEFAULT_LANGUAGE        | Default language to be used                                                                                                                        | en                                     |
| GOOSER_MAIL_FROM               | The mail address from which mails will be sent by the server                                                                                       | the value from GOOSER_SMTP_USERNAME    |
| GOOSER_MONGO_DB                | Name of the mongodb database                                                                                                                       | db                                     |
| GOOSER_MONGO_GROUPS_COLLECTION | Name of the mongodb groups collection                                                                                                              | groups                                 |
| GOOSER_MONGO_URL               | Url for the mongodb connection                                                                                                                     | mongodb://localhost:27017              |
| GOOSER_MONGO_USERS_COLLECTION  | Name of the mongodb users collection                                                                                                               | users                                  |
| GOOSER_OAUTH_URL               | Base url for oauth (will be used to query /userinfo)                                                                                               | http://localhost:4444                  |
| GOOSER_PORT                    | Port on which the server should be run                                                                                                             | 50051                                  |
| GOOSER_RESET_PASSWORD_URL      | Base url for resetting passwords                                                                                                                   | http://localhost:1234/#/reset-password |
| GOOSER_SECRET                  | Secret used for encryption. Make sure to set this variable in production!                                                                          |                                        |
| GOOSER_SITE_NAME               | Site name used in mails                                                                                                                            | gooser                                 |
| GOOSER_SMTP_HOST               | Hostname for the smtp connection. If not defined, mails will be written to stdout.                                                                 |                                        |
| GOOSER_SMTP_PASSWORD           | Password for the smtp connection                                                                                                                   |                                        |
| GOOSER_SMTP_PORT               | Port for the smtp connection                                                                                                                       | 587                                    |
| GOOSER_SMTP_USERNAME           | Username for the smtp connection                                                                                                                   |                                        |

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
## [0.2.2] - 2020-08-23
### Fixed
* fix UTF8 subject when sending mail to confirm mail address

## [0.2.1] - 2020-08-23
### Fixed
* add timezone data and ca-certs to docker image

## [0.2.0] - 2020-06-03
### Changed
* improved pagination which considers if documents were created or deleted
### Fixed
* code & tests cleanup

## [0.1.4] - 2020-06-03
### Fixed
* fixed counting documents / total size

## [0.1.3] - 2020-06-02
### Fixed
* fixed Dockerfile PORT

## [0.1.2] - 2020-06-02
### Fixed
* fixed localization

## [0.1.1] - 2020-06-02
### Fixed
* moved Dockerfile to correct location

## [0.1.0] - 2020-06-01
### Added
* basic functionality to CRUD users & groups
* functionality to confirm mail addresses
* functionality to reset password
* RBAC

cmd/gooser-server/main.go
package main

import (
	"context"
	"log"
	"os"
	"os/signal"
	"syscall"

	"github.com/rbicker/gooser/internal/auth"
	"github.com/rbicker/gooser/internal/mailer"
	"github.com/rbicker/gooser/internal/server"
	"github.com/rbicker/gooser/internal/store"
	_ "github.com/rbicker/gooser/internal/translations"
	"github.com/rbicker/gooser/internal/utils"
)

func main() {
	infoLogger := log.New(os.Stdout, "INFO: ", log.Lmsgprefix+log.LstdFlags)
	errLogger := log.New(os.Stderr, "ERROR: ", log.Lmsgprefix+log.LstdFlags)
	var dbOpts []func(*store.MGO) error
	var srvOpts []func(*server.Server) error
	// get or create secret
	var secret string
	if s, ok := os.LookupEnv("GOOSER_SECRET"); ok {
		secret = s
	} else {
		infoLogger.Printf("because no secret was given, a random string will be used")
		infoLogger.Printf("this means things like password reset links won't survive a restart of the application")
		infoLogger.Printf("make sure to set the GOOSER_SECRET environment variable in production")
		secret = utils.RandomString(20)
	}
	// init db connection
	mongoUrl := utils.LookupEnv("GOOSER_MONGO_URL", "mongodb://localhost:27017")
	dbOpts = append(dbOpts, store.WithURL(mongoUrl))
	dbName := utils.LookupEnv("GOOSER_MONGO_DB", "db")
	dbOpts = append(dbOpts, store.WithDBName(dbName))
	usersColName := utils.LookupEnv("GOOSER_MONGO_USERS_COLLECTION", "users")
	dbOpts = append(dbOpts, store.WithUsersCollectionName(usersColName))
	groupsColName := utils.LookupEnv("GOOSER_MONGO_GROUPS_COLLECTION", "groups")
	dbOpts = append(dbOpts, store.WithGroupsCollectionName(groupsColName))
	db, err := store.NewMongoConnection(secret, dbOpts...)
	if err != nil {
		errLogger.Fatalf("unable to create mongodb connection: %s", err)
	}
	err = db.Connect()
	if err != nil {
		errLogger.Fatalf("unable to connect to mongodb: %s", err)
	}
	infoLogger.Println("connected to mongodb")
	// mailer
	var mailClient mailer.MailClient
	smtpHost, ok := os.LookupEnv("GOOSER_SMTP_HOST")
	if ok {
		smtpPort := utils.LookupEnv("GOOSER_SMTP_PORT", "587")
		smtpUsername, _ := os.LookupEnv("GOOSER_SMTP_USERNAME")
		smtpPassword, _ := os.LookupEnv("GOOSER_SMTP_PASSWORD")
		mailClient, err = mailer.NewTLSMailer(smtpHost, smtpPort, smtpUsername, smtpPassword)
		if err != nil {
			log.Fatalf("error while creating mail client: %s", err)
		}
	} else {
		infoLogger.Println("no SMTP settings given, sending mails by logging them to stdout")
		infoLogger.Println("to send real mails, have a look at the GOOSER_SMTP_* environment variables")
		mailClient = mailer.NewLogMailer(infoLogger)
	}
	mailFrom, _ := os.LookupEnv("GOOSER_MAIL_FROM")
	siteName := utils.LookupEnv("GOOSER_SITE_NAME", "gooser")
	confirmUrl := utils.LookupEnv("GOOSER_CONFIRM_URL", "http://localhost:1234/#/confirm-mail")
	resetPasswordUrl := utils.LookupEnv("GOOSER_RESET_PASSWORD_URL", "http://localhost:1234/#/reset-password")
	mailer, err := mailer.NewMailer(mailClient, mailFrom, siteName, confirmUrl, resetPasswordUrl)
	if err != nil {
		log.Fatalf("error while creating mailer: %s", err)
	}
	// init server
	srvOpts = append(srvOpts, server.EnableReflection())
	p := utils.LookupEnv("GOOSER_PORT", "50051")
	srvOpts = append(srvOpts, server.SetPort(p))
	oauthUrl := utils.LookupEnv("GOOSER_OAUTH_URL", "http://localhost:4444")
	oAuth, err := auth.NewOAuthClient(oauthUrl)
	if err != nil {
		errLogger.Fatalf("unable to create oAuth client: %s", err)
	}
	srv, err := server.NewServer(secret, db, oAuth, mailer, srvOpts...)
	if err != nil {
		errLogger.Fatalf("unable to create new gooser server: %s", err)
	}
	// init collections
	err = srv.InitCollections(context.Background())
	if err != nil {
		errLogger.Fatalf("unable to initialize collections: %s", err)
	}
	// channels
	errChan := make(chan error)
	stopChan := make(chan os.Signal)
	// bind OS events to the signal channel
	signal.Notify(stopChan, syscall.SIGTERM, syscall.SIGINT)
	// serve in a go routine
	go func() {
		infoLogger.Println("starting gooser server")
		if err := srv.Serve(); err != nil {
			errChan <- err
		}
	}()
	// terminate gracefully before leaving the main function
	defer func() {
		infoLogger.Println("stopping grpc server")
		srv.Stop()
		infoLogger.Println("disconnecting from mongodb")
		err := db.Disconnect(context.TODO())
		if err != nil {
			errLogger.Fatalf("error while disconnecting from mongodb: %s", err)
		}
	}()
	// block until either OS signal, or server fatal error
	select {
	case err := <-errChan:
		errLogger.Printf("Fatal error: %v\n", err)
	case <-stopChan:
	}
}
